File: Leetcode/Python/3903.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)
        suffix = [float("inf")] * n
        suffix[-1] = nums[-1]
        for i in range(len(nums) - 2, -1, -1):
            suffix[i] = min(suffix[i + 1], nums[i])
        dist = float("inf")
        cmax = nums[0]
        for i in range(n):
            cmax = max(cmax, nums[i])
            cdist = abs(cmax - suffix[i])
            if dist > cdist:
                dist = cdist
                if dist <= k:
                    return i
        return -1

        

File: Leetcode/cpp/3903.cpp
class Solution {
public:
    int firstStableIndex(vector<int>& nums, int k) {
        int n = nums.size();
        std::vector<int> suffix(n, 0);
        suffix[n - 1] = nums.back();
        for (int i = n - 2; i >= 0; --i) {
            suffix[i] = std::min<int>(suffix[i + 1], nums[i]);
        }
        int cmax = nums[0];
        for (int j = 0; j < n; ++j) {
            cmax = std::max<int>(cmax, nums[j]);
            int dist = cmax - suffix[j];
            if (dist <= k) {
                return j;
            }
        }
        return -1;
    }
};
